### ISSUE TITLE
Language service infers incorrect type in structural directive using "as"

### ISSUE THREAD

**Issue by Airblader** [angular/vscode-ng-language-service#651]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/language-service.

### Is this a regression?

I don't think so.

### Description

According to the [docs](https://angular.io/guide/structural-directives), the structural directive syntax

```
<ng-container *appTest="exp as value"></ng-container>
```

is desuggared into the following:

```
<ng-template [appTest]="exp" let-value="appTest">
    <ng-container></ng-container>
</ng-template>
```

If the structural directive has an `appTest` property on the context with which it creates the template which has a different type than what the `@Input() appTest` input takes (the type of `exp`), everything works functionally, but the language service seems to infer the type of `value` to be that of `exp` rather than the type defined by the context.

## 🔬 Minimal Reproduction

```
export class TestContext {
    public $implicit: number;
    public appTest: string;
}

@Directive({
    selector: "[appTest]",
})
export class AppTestDirective implements OnInit {
    exp: number;

    constructor (private viewContainer: ViewContainerRef, private templateRef: TemplateRef<any>) {}

    ngOnInit() {
        const context = new TestContext();
        context.$implicit = this.exp;
        context.appTest = `Test: ${this.exp}`;
        this.viewContainer.createEmbeddedView(this.templateRef, context);
    }

    @Input()
    public set appTest(exp: number) {
        this.exp = exp;
    }
}
```

Usage:

```
<p *appTest="5 as value">{{ value }}</p>
```

This *will* print `Test: 5`, but the inferred type for `value` is `number`.

It can be argued whether `x as y` should allow that `x` and `y` are of different types since this can be counter-intuitive (though in my use-case it probably isn't), but either way the behavior should be consistent.

## 🌍  Your Environment

**Angular Version:**
<pre><code>8.2.13</code></pre>

**Comment by ayazhafiz** [comment]
The language service can only determine the type of the template context from the type of the `templateRef` defined on the component. Maybe this is not the best way (a better way could be to read the return type of `ngOnInit`), but I think this is the only way we can do it until the language service is flipped to use the Ivy compiler as a backend.

If you update `private templateRef: TemplateRef<any>` to be `private templateRef: TemplateRef<TestContext>`, you will see the behavior you expect.

**Comment by Airblader** [comment]
@ayazhafiz In my case the template ref is typed accordingly but the issue is there nonetheless.

**Comment by ayazhafiz** [comment]
> @ayazhafiz In my case the template ref is typed accordingly but the issue is there nonetheless.

You mean if you use `private templateRef: TemplateRef<TestContext>` the issue persists? That's definitely unexpected even with today's language service... I'll take a look.

**Comment by Airblader** [comment]
That's how my directive is set up and I've just had a user report this issue again. See the linked issue. You can find the directive itself here: https://github.com/TNG/ngx-structurals/blob/master/projects/ngx-structurals/src/lib/ngx-subscribe/ngx-subscribe.directive.ts

**Comment by ayazhafiz** [comment]
Hm.. in your directive's case the input (`ngxSubscribe` and `ngxSubscribeOf`) is defined on the directive but not on the [template context](https://github.com/TNG/ngx-structurals/blob/master/projects/ngx-structurals/src/lib/ngx-subscribe/ngx-subscribe.directive.ts#L9). The language service should handle this (I'll look into fixing it later today), but this issue is a little bit different than the original case of the language service not picking up the template context correctly.

FWIW, the linked issue with your repo refers to a compile time error and is unrelated to the language service.

**Comment by kyliau** [comment]
This has been fixed by the new Ivy-native language service, released in [v11.1.0](https://github.com/angular/vscode-ng-language-service/releases/tag/v11.1.0).
It's an opt-in feature for now, please give it a try and let us know if you have any feedback.
Please make sure your project has [`strictTemplates`](https://angular.io/guide/angular-compiler-options#stricttemplates) enabled in `angularCompilerOptions`.
In case you run into similar bug, please file a new issue. I'll close this for now.
